enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,320 @@
+  known firmware settings.
+
+- As an operator of hardware that does not support virtual media, I want to
+  deploy OpenShift nodes using iPXE network boot.
+
+- As an operator, I want OpenShift to ensure that the deployment ISO is
+  correctly disconnected on installation finish and cannot be accidentally
+  booted into.
+
+### API Extensions
+
+All required API extensions are already present.
+
+A new annotation will be recognized on `PreprovisioningImage` resources:
+`baremetal.openshift.io/ignition-override-uri`.
+
+### Implementation Details/Notes/Constraints
+
+#### Coordination between agents
+
+The first agent to start will be the Ironic agent since it is responsible
+for coordinating the whole flow and (indirectly through Ironic) updating the
+`BareMetalHost`'s status. *Inspection* and additional configuration (RAID etc)
+will be handled by the Ironic agent alone.
+
+BMAC will be able to stop synchronizing the inventory to `BareMetalHost`
+objects since this task will be handled by Ironic and baremetal-operator.
+
+Currently, the provisioning is initiated by BMAC (the assisted component
+responsible for interactions with baremetal-operator) by applying the following
+change to the `BareMetalHost`'s spec:
+
+```yaml
+Image:
+  URL: url/of/assisted/agent.iso
+  DiskFormat: live-iso
+```
+
+After this enhancement, BMAC will instead apply:
+
+```yaml
+CustomDeploy:
+  Method: install_assisted
+```
+
+This instructs Ironic to find and execute a *custom deploy step* called
+`install_assisted`, which will be provided by an OpenShift specific Ironic
+agent plugin. This step will use D-Bus to order systemd to start the assisted
+agent and watch its status.
+
+If the assisted agent exits successfully, the Ironic agent will report success
+to Ironic and baremetal-operator. Otherwise, a failure will be recorded.
+
+The `CustomDeploy` field can be set from the very beginning as well. Then BMO
+will conduct inspection, apply firmware settings (if any) and prepare for
+the installation immediately.
+
+### Minimal ISO
+
+ICC is currently not capable of building a *mimimal ISO*, i.e. a CoreOS ISO
+without the root filesystems. While not a hard blocker of this work, the
+minimal ISO support should be implemented because
+- it improves compatibility with BMCs that do not accept large ISOs,
+- it reduces the resource footprint on the Ironic side.
+
+#### Notes
+
+- The new annotation with accept a URI instead of a full Ignition configuration
+  or a link to a secret because the configuration can be quite large when
+  many hosts are present.
+
+- ICC will replicate the graceful wait period logic from the assisted service:
+  the image will be built not earlier than 60 seconds from the last change
+  to the `PreprovisioningImage` object. This is done for two reasons: to give
+  the user enough time to apply all changes that may affect the build and to
+  reduce the probability of a race between BMAC and ICC.
```

Comment:
Tracking modification time on CRs can be tricky. Is this really necessary? If so, let's describe how to track the update time.

Reply:
Aha, assisted is using `Status.CreatedTime`, not modification time. Which is, indeed, a different thing, but fits this use case reasonably well.